wd.c

```diff
@@ -604,13 +604,13 @@ struct uacce_dev *wd_get_accel_dev(char *alg_name)
 	if (!list)
 		return NULL;
 
-	dev = list->dev;
-
 	while (list) {
 		tmp = numa_distance(node, list->dev->numa_id);
 		if (dis > tmp && tmp > 0) {
-			dev = list->dev;
-			dis = tmp;
+			if (wd_get_avail_ctx(list->dev) > 0) {
```

Comment:
request_ctx will handle the fail case, here only return the best dev

Reply:
I think there's a potential issue. User expects to allocate a right device, but may get a device with very limited contexts. If user requests contexts failure, what should he do next?

Reply:
If so, it's unnecessary to check avail context at here. Even process A requests a device with enough contexts. It may be switched out, and process B and C allocate all of contexts. When process A switches back, there're no enough contexts already. Process A has to wait other release contexts.

Reply:
current usage is request_ctx just before doing the work, inside the process, not prepare resources and then fork, it is not supported by uadk

Reply:
It doesn't matter. I said processes. In a system, there're always lots of processes running. Switching process is controlled by OS, not user.

Reply:
Currently you're proposing there's at least one context in the device. But there's no difference between no contexts and one context in the real system. Since any other process running in the system could allocate the left context before you.
You're just allocating device, not preserving contexts in the device.
So I think either not considering contexts or consider contexts balance in devices.